File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
            
        q = deque()
        q.append(root)
        ans = []
        switch = 0
        while q:
            temp_res = []
            for _ in range(len(q)):
                curr = q.popleft()
                temp_res.append(curr.val)

                if curr.left is not None:
                    q.append(curr.left)
                    
                if curr.right is not None:
                    q.append(curr.right)
                    
            if switch == 0:
                ans.append(temp_res)
            else:
                temp_res.reverse()
                ans.append(temp_res)
                
            switch = switch ^ 1
            
        return ans

        

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> levelOrder = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean isEven = true;
        while (!queue.isEmpty()) {
            int size = queue.size();
            levelOrder =  new ArrayList<>();
            for(int i = 0; i < size; i++) {
                TreeNode currentNode = queue.poll();
                if(!isEven){
                   levelOrder.add(0, currentNode.val);
                }else{
                    levelOrder.add(currentNode.val);
                }
                if (currentNode.left != null) {
                        queue.offer(currentNode.left);
                }
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }
            
            isEven =!isEven;
            result.add(levelOrder);
        }
        return result;
    }
}